### ISSUE TITLE
Handling of legacy identifier URIs.

### ISSUE THREAD

**Issue by nbaju1** [hashicorp/terraform-provider-azuread#894]
In our tenant we have multiple applications with legacy identifier URIs, i.e. URIs that, if removed, can't be reapplied to the application by using the API due to lacking host name, unverified domain etc. The apps are functioning with these URIs.

The problem is when we try to manage these apps through Terraform since both validation and apply fails when the config file contains legacy URIs. Is there a to make Terraform ignore identifier_uris when the config and app is in sync, or some other way to handle this issue?

I realize that this is a bit of a open-ended question, but I've tried so many different hacks to make this work and feel like I'm out of options. Would rather not update the identifier URIs for 300+ applications.

Ref legacy identifier URIs: https://learn.microsoft.com/en-us/azure/active-directory/develop/reference-breaking-changes

**Comment by manicminer** [comment]
Hi @nbaju1, thanks for raising this query. Unfortunately it will not be possible to manage these identifier URIs with Terraform. As you allude to, any changes that Terraform may try to effect will incur API validation anyway; the validation within Terraform is intended to meet the same validation rules.

You may be able to manage these applications by adding `identifier_uris` to the `ignore_changes` lifecycle argument, whilst ensuring the `identifier_uris` property is omitted for these same applications.

**Comment by nbaju1** [comment]
Thanks for the reply, @manicminer. Weirdly enough, your suggestion works for some apps, but not for others.
For all apps I remove the identifier URIs from the config file prior to terraform validate, plan and apply. Any changes to the identifier URIs is done via the Graph API post apply. `identifier_uris` is added to `ignore_changes` on the application.

The following app had 4 identifier URIs. Apply went without issues. Apply fails for this app if the URIs are not removed and `identifier_uris` are not added to `ignore_changes`.

```
Terraform will perform the following actions:

  # azuread_application.<redacted>_application will be updated in-place
  ~ resource "azuread_application" "<redacted>_application" {
        id                             = "<redacted>"
      ~ owners                         = [
          - "<redacted>",
            # (1 unchanged element hidden)
        ]
        tags                           = [
            "<redacted>",
        ]
        # (14 unchanged attributes hidden)

        # (9 unchanged blocks hidden)
    }

  # azuread_service_principal.<redacted>_service_principal will be updated in-place
  ~ resource "azuread_service_principal" "<redacted>_service_principal" {
        id                            = "<redacted>"
      ~ owners                        = [
          - "<redacted>",
            # (1 unchanged element hidden)
        ]
      - preferred_single_sign_on_mode = "saml" -> null
        tags                          = [
            "HideApp",
            "WindowsAzureActiveDirectoryCustomSingleSignOnApplication",
            "WindowsAzureActiveDirectoryGalleryApplicationNonPrimaryV1",
            "WindowsAzureActiveDirectoryIntegratedApp",
            "<redacted>",
        ]
        # (20 unchanged attributes hidden)

        # (4 unchanged blocks hidden)
    }

Plan: 0 to add, 2 to change, 0 to destroy
```

This next app had 1 identifier URI, but failed on apply due to unverified domain.

```
Terraform will perform the following actions:

  # azuread_application.<redacted>_application will be updated in-place
  ~ resource "azuread_application" "<redacted>_application" {
        id                             = "<redacted>"
      ~ owners                         = [
          + "<redacted>",
        ]
      ~ tags                           = [
          + "<redacted>",
        ]
        # (14 unchanged attributes hidden)

        # (8 unchanged blocks hidden)
    }

  # azuread_service_principal.<redacted>_service_principal will be updated in-place
  ~ resource "azuread_service_principal" "<redacted>_service_principal" {
      + description                  = "Service principal corresponding to application '<redacted>'."
        id                           = "<redacted>"
      ~ owners                       = [
          + "<redacted>",
        ]
      ~ tags                         = [
          - "<redacted>",
          + "WindowsAzureActiveDirectoryCustomSingleSignOnApplication",
          + "<redacted>",
            # (2 unchanged elements hidden)
        ]
      + use_existing                 = true
        # (17 unchanged attributes hidden)

        # (3 unchanged blocks hidden)
    }

Plan: 0 to add, 2 to change, 0 to destroy.
```

This seems inconsistent?

**Comment by manicminer** [comment]
@nbaju1 This is probably inconsistent API behavior - for both those cases the payload would include the `identifierUris` field with the current value (e.g. a no-op), but I guess the API is sometimes ok with that and sometimes not.

In future we will be trying to improve the logic here to only send fields in patch requests that have actually changed, or that must be included to maintain their value. I don't think identifierUris is one of the latter but at the moment we are including that in the body.

**Comment by nbaju1** [comment]
I found the culprit that caused the inconsistent behavior. For the apps that didn't work, `requestedAccessTokenVersion` was set to `null` on the app. Doing a patch request on the app to update this to "1" results in the failure above, even if `identifierUris` is not present in the patch payload. I see in the code that you default this value to "1", and I assume this is then present in the patch payload. 

I removed the troublesome identifier URI, patched `requestedAccessTokenVersion` to "1", re-added the URI in the SAML settings page (no validation here, can put anything) an re-ran the terraform apply pipeline. It ran without errors.
